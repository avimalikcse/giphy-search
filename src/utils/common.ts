
/**
 * Method to prevent repeated execution of a particular function within a certain time range
 * E.G: prevent multiple API calls if use click on search button multiple time
 * 
 * @param {Function} fn function that needs to be throttled 
 * @param {number} wait time it should wait before call
 * @returns void
 */

export const throttle = (fn: any, wait: number = 300) => {
  let inThrottle: boolean,
    lastFn: ReturnType<typeof setTimeout>,
    lastTime: number;
  return function (this: any) {
    const context = this,
      args = arguments;
    if (!inThrottle) {
      fn.apply(context, args);
      lastTime = Date.now();
      inThrottle = true;
    } else {
      clearTimeout(lastFn);
      lastFn = setTimeout(() => {
        if (Date.now() - lastTime >= wait) {
          fn.apply(context, args);
          lastTime = Date.now();
        }
      }, Math.max(wait - (Date.now() - lastTime), 0));
    }
  };
};
