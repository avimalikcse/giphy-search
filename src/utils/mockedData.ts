/**
 * File to keep mocked data
 */
export const MOCKED_DATA =
{

    "id": "9JgeSP0jlRAVBOG9FD",
    "title": "Bill Nye Spinning GIF by Nike",
    "rating": "g",
    "content_url": "",
    "source_tld": "",
    "source_post_url": "",
    "is_sticker": 0,
    "import_datetime": "2018-03-02 20:36:17",
    "trending_datetime": "2018-03-05 04:20:07",
    "images": {
        "original": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif",
            "mp4_size": "159778",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.mp4",
            "webp_size": "356696",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.webp",
            "frames": "33",
            "hash": "65151deca02da1bfa8c196a328e5115b"
        },
        "downsized": {
            "height": "270",
            "width": "480",
            "size": "1110011",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized.gif"
        },
        "downsized_large": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif"
        },
        "downsized_medium": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.gif"
        },
        "downsized_small": {
            "height": "228",
            "width": "405",
            "mp4_size": "54997",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized-small.mp4"
        },
        "downsized_still": {
            "height": "270",
            "width": "480",
            "size": "34777",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-downsized_s.gif"
        },
        "fixed_height": {
            "height": "200",
            "width": "356",
            "size": "814594",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.gif",
            "mp4_size": "88466",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.mp4",
            "webp_size": "296408",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200.webp"
        },
        "fixed_height_downsampled": {
            "height": "200",
            "width": "356",
            "size": "165256",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_d.gif",
            "webp_size": "100664",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_d.webp"
        },
        "fixed_height_small": {
            "height": "100",
            "width": "178",
            "size": "269203",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.gif",
            "mp4_size": "30509",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.mp4",
            "webp_size": "100108",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100.webp"
        },
        "fixed_height_small_still": {
            "height": "100",
            "width": "178",
            "size": "10714",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100_s.gif"
        },
        "fixed_height_still": {
            "height": "200",
            "width": "356",
            "size": "25582",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200_s.gif"
        },
        "fixed_width": {
            "height": "113",
            "width": "200",
            "size": "327748",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.gif",
            "mp4_size": "36086",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.mp4",
            "webp_size": "121676",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w.webp"
        },
        "fixed_width_downsampled": {
            "height": "113",
            "width": "200",
            "size": "63820",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_d.gif",
            "webp_size": "39244",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_d.webp"
        },
        "fixed_width_small": {
            "height": "57",
            "width": "100",
            "size": "105544",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.gif",
            "mp4_size": "13101",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.mp4",
            "webp_size": "42322",
            "webp": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w.webp"
        },
        "fixed_width_small_still": {
            "height": "57",
            "width": "100",
            "size": "4120",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/100w_s.gif"
        },
        "fixed_width_still": {
            "height": "113",
            "width": "200",
            "size": "13047",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/200w_s.gif"
        },
        "looping": {
            "mp4_size": "2658625",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-loop.mp4"
        },
        "original_still": {
            "height": "270",
            "width": "480",
            "size": "90732",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif"
        },
        "original_mp4": {
            "height": "270",
            "width": "480",
            "mp4_size": "159778",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy.mp4"
        },
        "preview": {
            "height": "122",
            "width": "216",
            "mp4_size": "19345",
            "mp4": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.mp4"
        },
        "preview_gif": {
            "height": "80",
            "width": "142",
            "size": "48869",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.gif"
        },
        "preview_webp": {
            "height": "108",
            "width": "192",
            "size": "43254",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy-preview.webp"
        },
        "480w_still": {
            "height": "270",
            "width": "480",
            "size": "2021711",
            "url": "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/480w_s.jpg"
        }
    },
    "user": {
        "avatar_url": "https://media4.giphy.com/avatars/nike/gafTSy1Qou4l.png",
        "banner_image": "https://media4.giphy.com/headers/nike/mPbbY4srw3HC.png",
        "banner_url": "https://media4.giphy.com/headers/nike/mPbbY4srw3HC.png",
        "profile_url": "https://giphy.com/nike/",
        "username": "nike",
        "display_name": "Nike",
        "description": "Just Do it",
        "instagram_url": "https://instagram.com/Nike",
        "website_url": "http://www.nike.com/justdoit",
        "is_verified": true
    },
    "analytics_response_payload": "e=Z2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n",
    "analytics": {
        "onload": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=SEEN"
        },
        "onclick": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=CLICK"
        },
        "onsent": {
            "url": "https://giphy-analytics.giphy.com/v2/pingback_simple?analytics_response_payload=e%3DZ2lmX2lkPTlKZ2VTUDBqbFJBVkJPRzlGRCZldmVudF90eXBlPUdJRl9TRUFSQ0gmY2lkPTQ2ZDdjMmI4cjg0MGx2YXRveTAzdmt0cjB3aGoyenI2dWdwczhhdXY0MXh1djJnayZjdD1n&action_type=SENT"
        }
    }
}

export const MOCKED_TYPE_DATA =
{
    id: "9JgeSP0jlRAVBOG9FD",
    title: "Bill Nye Spinning GIF by Nike",
    thumbnail: "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif",
    gifURL: "https://media2.giphy.com/media/9JgeSP0jlRAVBOG9FD/giphy_s.gif",
    width: 400,
    height: 200,
}