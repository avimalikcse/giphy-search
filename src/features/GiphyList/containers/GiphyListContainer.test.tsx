import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MOCKED_DATA } from '../../../utils/mockedData';
import GiphyListingContainer from './GiphyListContainer';



global.fetch = jest.fn(() =>
    Promise.resolve({
        json: () =>
            Promise.resolve({
                status: 200,
                data: { data: [MOCKED_DATA], pagination: { total: 1, limit: 10, offset: 0 } },
            }),
    }),
) as jest.Mock;

test('Test if Search Bar is rendered ', async () => {
    render(< GiphyListingContainer />);
    await waitFor(() => {
        const searchBox = screen.getByPlaceholderText(/search GIFs/i);
        expect(searchBox).toBeInTheDocument();
    })
});


test('Test snapshots of the components', async () => {
    const { asFragment } = render(< GiphyListingContainer />);
    await waitFor(() => {
        expect(asFragment()).toMatchSnapshot();
    })
});
