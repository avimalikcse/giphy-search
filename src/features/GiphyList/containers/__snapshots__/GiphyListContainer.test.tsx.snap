// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshots of the components 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-4cgrtw-MuiPaper-root"
  >
    <form
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1rha8zw-MuiPaper-root"
    >
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-piqts5-MuiInputBase-root"
      >
        <input
          aria-label="search GIFs"
          class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
          placeholder="search GIFs"
          type="text"
          value=""
        />
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-jbezd2-MuiDivider-root"
      />
      <button
        aria-label="search"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1gtpq4r-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
  <div
    class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-18x8o3k-MuiTypography-root"
  >
    Please Search a GIF
  </div>
</DocumentFragment>
`;
